gnomad_qc/v3/subset/subset.py

```diff
@@ -0,0 +1,243 @@
+        if ft == hl.dtype("int64"):
+            ht = ht.annotate(
+                info=ht.info.annotate(**{f: hl.int32(hl.min(2 ** 31 - 1, ht.info[f]))})
+            )
+        elif ft == hl.dtype("float64"):
+            ht = ht.annotate(
+                info=ht.info.annotate(
+                    **{f: hl.float32(hl.min(2 ** 31 - 1, ht.info[f]))}
+                )
+            )
+        elif ft == hl.dtype("array<int64>"):
+            ht = ht.annotate(
+                info=ht.info.annotate(
+                    **{f: ht.info[f].map(lambda x: hl.int32(hl.min(2 ** 31 - 1, x)))}
+                )
+            )
+        elif ft == hl.dtype("array<float64>"):
+            ht = ht.annotate(
+                info=ht.info.annotate(
+                    **{f: ht.info[f].map(lambda x: hl.float32(hl.min(2 ** 31 - 1, x)))}
+                )
+            )
+    return ht
+
+
+def compute_partitions(mt, entry_size=3.5, partition_size=128000000) -> int:
+    """Computes a very rough estimate for the optimal number of partitions in a MT
+     using a the hail recommended partition size of 128MB and the rough estimate 
+     of 3.5B per entry in the gnomAD sparse MT.
+
+    :param mt: MatrixTable that will be resized
+    :param entry_size: Average size in bytes that a single entry requires , defaults to 3.5
+    :param partition_size: Amount of data per partition. Hail recommends 128MB, defaults to 128000000
+    :return: number of partitions for resizing MT  
+    """
+    rows, columns = mt.count()
+    mt_disk_est = rows * columns * entry_size
+    n_partitions = (
+        int(mt_disk_est / partition_size) if (mt_disk_est / partition_size) > 20 else 20
+    )
+    return n_partitions
+
+
+def main(args):
+    hl.init(
+        log="/subset.log", tmp_dir="hdfs:///subset.tmp/", default_reference="GRCh38"
+    )
+    pop = args.pop
+    subset = args.subset
+    mt = get_gnomad_v3_mt(
+        key_by_locus_and_alleles=True, remove_hard_filtered_samples=False
+    )
+    info_ht = get_info().ht()
+    info_ht = format_info_for_vcf(info_ht)
+
+    if "SB" in info_ht.info and not isinstance(
+        info_ht.info.SB, hl.expr.ArrayNumericExpression
+    ):
+        info_ht = info_ht.annotate(
+            info=info_ht.info.annotate(SB=hl.flatten(info_ht.info.SB))
+        )
+
+    meta = metadata.ht()
+
+    if pop and pop in GENOME_POPS:
+        logger.info(f"Subsetting samples to {pop} population")
+        mt = mt.annotate_cols(pop=meta[mt.col_key].pop)
+        mt = mt.filter_cols(mt.pop == pop)
+        mt = mt.cols().drop("pop")
+
+    if subset:
+        mt = subset_samples_and_variants(mt, subset, sparse=True, gt_expr="LGT")
+
+    if args.pull_meta:
+        meta = meta.semi_join(mt.cols())
+        meta.export(f"{args.output_path}/metadata.tsv.bgz")
+
+    mt = hl.experimental.sparse_split_multi(mt, filter_changed_loci=True)
+    mt = hl.experimental.densify(mt)
+
+    mt = mt.filter_rows(hl.agg.any(mt.GT.is_non_ref()))
```

Comment:
These two lines should be switched because `mt = mt.filter_rows(hl.len(mt.alleles) > 1)` is less costly than `mt = mt.filter_rows(hl.agg.any(mt.GT.is_non_ref()))` since it doesn't operate on the entries. In most programming langs, when evaluating logical statements with multiple conditions (which these will be collapsed into (`mt = mt.filter_rows(hl.agg.any(mt.GT.is_non_ref()) & (hl.len(mt.alleles) > 1))`), the cheaper (or most common, depending on the situation) should always come first as oftentimes the last part of the statement is only evaluated as needed (in this case, `hl.agg.any(mt.GT.is_non_ref())` would only run iff `hl.len(mt.alleles) > 1)`. I asked Tim on Zulip to be sure that hail is sensitive to ordering and he said we should switch (without providing much details though). In any case, I think it's a good habit to take to think about this and there is no cost associated with ordering them based on their cost.

Reply:
This makes complete sense